xfr_test.go

```diff
@@ -0,0 +1,195 @@
+
+	s, addrstr, err := RunLocalTCPServer(":0")
+	if err != nil {
+		t.Fatalf("unable to run test server: %s", err)
+	}
+	defer s.Shutdown()
+
+	tr := new(Transfer)
+	m := new(Msg)
+	m.SetAxfr("miek.nl.")
+
+	c, err := tr.In(m, addrstr)
+	if err != nil {
+		t.Fatal("failed to zone transfer in", err)
+	}
+
+	for msg := range c {
+		if msg.Error == nil {
+			t.Fatal("failed to catch 'no SOA' error")
+		}
+	}
+}
+
+func TestSingleEnvelopeXfr(t *testing.T) {
+	HandleFunc("miek.nl.", SingleEnvelopeXfrServer)
+	defer HandleRemove("miek.nl.")
+
+	s, addrstr, err := RunLocalTCPServerWithTsig(":0", tsigSecret)
+	if err != nil {
+		t.Fatalf("unable to run test server: %s", err)
+	}
+	defer s.Shutdown()
+
+	testCases := []struct {
+		name string
+		tsig map[string]string
+	}{
+		{"empty", nil},
+		{"valid", tsigSecret},
+	}
+
+	for _, tc := range testCases {
+		t.Run(fmt.Sprintf("Testing %v TSIG", tc.name), axfrTestingSuite(addrstr))
+	}
+}
+
+func TestMultiEnvelopeXfr(t *testing.T) {
+	HandleFunc("miek.nl.", MultipleEnvelopeXfrServer)
+	defer HandleRemove("miek.nl.")
+
+	s, addrstr, err := RunLocalTCPServerWithTsig(":0", tsigSecret)
+	if err != nil {
+		t.Fatalf("unable to run test server: %s", err)
+	}
+	defer s.Shutdown()
+
+	testCases := []struct {
+		name string
+		tsig map[string]string
+	}{
+		{"empty", nil},
+		{"valid", tsigSecret},
+	}
+
+	for _, tc := range testCases {
+		t.Run(fmt.Sprintf("Testing %v TSIG", tc.name), axfrTestingSuite(addrstr))
+	}
+}
+
+func RunLocalTCPServerWithTsig(laddr string, tsig map[string]string) (*Server, string, error) {
+	server, l, _, err := RunLocalTCPServerWithFinChanWithTsig(laddr, tsig)
+
+	return server, l, err
+}
+
+func RunLocalTCPServerWithFinChanWithTsig(laddr string, tsig map[string]string) (*Server, string, chan error, error) {
+	l, err := net.Listen("tcp", laddr)
+	if err != nil {
+		return nil, "", nil, err
+	}
+
+	server := &Server{Listener: l, ReadTimeout: time.Hour, WriteTimeout: time.Hour, TsigSecret: tsig}
+
+	waitLock := sync.Mutex{}
+	waitLock.Lock()
+	server.NotifyStartedFunc = waitLock.Unlock
+
+	// See the comment in RunLocalUDPServerWithFinChan as to
+	// why fin must be buffered.
+	fin := make(chan error, 1)
+
+	go func() {
+		fin <- server.ActivateAndServe()
+		l.Close()
+	}()
+
+	waitLock.Lock()
+	return server, l.Addr().String(), fin, nil
+}
+
+func axfrTestingSuite(addrstr string) func(*testing.T) {
+	return func(t *testing.T) {
+		tr := new(Transfer)
+		m := new(Msg)
+		m.SetAxfr("miek.nl.")
+
+		c, err := tr.In(m, addrstr)
+		if err != nil {
+			t.Fatal("failed to zone transfer in", err)
+		}
+
+		var records []RR
+		for msg := range c {
+			if msg.Error != nil {
+				t.Fatal(msg.Error)
+			}
+			for _, rr := range msg.RR {
+				records = append(records, rr)
+			}
+		}
+
+		if len(records) != len(testData) {
+			t.Fatalf("bad axfr: expected %v, got %v", records, testData)
+		}
+
+		for i := range records {
+			if records[i].String() != testData[i].String() {
```

Comment:
I don’t think this is the proper way to compare RRs.

Reply:
I suspect not.  Would it make more sense to use `isDuplicate()` instead?

```
for i := range records {
	if !records[i].isDuplicate(xfrTestData[i]) {
		t.Fatalf("bad axfr: expected %v, got %v", records, xfrTestData)
	}
}
```

Reply:
Yes it would, but use the exported `IsDuplicate` function and not the I exported method.

Reply:
Thanks.  I wasn't tracking the exported version.